File: dynamic/dynamic/report/mode_of_payment_report/mode_of_payment_report.py
# Copyright (c) 2013, Dynamic Technology and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from erpnext import get_company_currency, get_default_company
def execute(filters=None):
	columns, data =get_columns(filters), get_data(filters)
	return columns, data


def get_balance(row, balance, debit_field, credit_field):
	balance += (row.get(debit_field, 0) -  row.get(credit_field, 0))

	return balance

def get_result_as_list(data, filters,oppening_balance=None):
	balance, balance_in_account_currency = 0, 0
	oppening_balance_total = 0
	if oppening_balance:
		for row in oppening_balance:
			oppening_balance_total += (row.get('debit', 0) -  row.get('credit', 0))
	
	row = {"balance":oppening_balance_total,"voucher_type":"Opening"}
	balance = oppening_balance_total or 0

	for d in data:
		if not d.get('posting_date'):
			balance, balance_in_account_currency = 0, 0

		balance = get_balance(d, balance, 'debit', 'credit')
		d['balance'] = balance

	data.insert(0,row)
	return data



def get_data(filters=None):
	oppening_balance = ''
	if filters.from_date > filters.to_date:
		frappe.throw(_("From Date must be before To Date"))
	gl_type=  filters.get("account_type") 
	accounts = f""" (select name from `tabAccount` WHERE account_type = '{gl_type}' )"""
	if filters.get("mode_of_payment") :
		accounts =f""" (SELECT default_account FROM 
							`tabMode of Payment Account` WHERE parent = '{filters.get("mode_of_payment")}' )"""
	
	date_qyery = f"""  AND posting_date between date('{filters.from_date }')  AND  date('{filters.to_date }') """

	base_query = f"""SELECT name as gl_entry , posting_date ,account , debit ,credit ,0 as balance  ,voucher_type 
				FROM `tabGL Entry`  
				 WHERE account in {accounts} 
	   """ 
	sql_qery = base_query+date_qyery + "ORDER BY posting_date"
	data = frappe.db.sql(sql_qery ,as_dict=1)

	if filters.get("mode_of_payment") :
		oppening_balance = base_query + f"""  AND posting_date < '{filters.from_date }' """ + "ORDER BY posting_date"
		oppening_balance = frappe.db.sql(oppening_balance ,as_dict=1)

	return (get_result_as_list(data , filters,oppening_balance) )

def get_columns(filters=None) :
	currency = get_company_currency(filters["company"]) if filters else get_default_company()
	columns = [
		{
			"label": _("GL Entry"),
			"fieldname": "gl_entry",
			"fieldtype": "Link",
			"options": "GL Entry",
			"hidden": 1
		},
		{
			"label": _("Posting Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width":130
		},
		{
			"label": _("Account"),
			"fieldname": "account",
			"fieldtype": "Link",
			"options": "Account",
			"width":270
		},
		{
			"label": _("Debit ({0})".format(currency)),
			"fieldname": "debit",
			"fieldtype": "Float",
			"width": 100
		},
		{
			"label": _("Credit ({0})".format(currency)),
			"fieldname": "credit",
			"fieldtype": "Float",
			"width": 100
		},
		{
			"label": _("Balance ({0})".format(currency)),
			"fieldname": "balance",
			"fieldtype": "Float",
			"width": 130
		},
		{
			"label": _("Voucher Type"),
			"fieldname": "voucher_type",
			"width": 180
		},
		{
			"label": _("Voucher No"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "voucher_type",
			"width": 180
		},
			
	]
	return columns

File: dynamic/dynamic/report/mode_of_payment_report/mode_of_payment_report.js


frappe.query_reports["Mode Of Payment Report"] = {
    onload: function(report){
		if(frappe.query_report.get_filter_value('make_hidden') == 1)
        {
            $('[data-fieldname=mode_of_payment]').attr("disabled", true);
        }
	},
	"filters": [
		{
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },

        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1,
            "width": "60px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1,
            "width": "60px"
        },
		{
            "fieldtype": "Break",
        },
		{
            "fieldname": "mode_of_payment",
            "label": __("Mode Of Payment"),
            "fieldtype": "Link",
            "options": "Mode of Payment",
            "reqd": 0
        },
        {
            "fieldname": "make_hidden",
            "label": __("make_hidden"),
            "fieldtype": "Select",
            "options": "0\n1",
            "default": "0",
            "hidden": "1",
            // on_change: function() {
            //     console.log("changed")
            //     console.log(frappe.query_report)
            //     // frappe.query_report.filters[3].readonly = 1
            //     $('[data-fieldname=mode_of_payment]').attr("disabled", true);
            //     frappe.query_report.refresh()
				
			// }
        },
		{
            "fieldname": "account_type",
            "label": __("Account Type"),
            "fieldtype": "Select",
            "options": "Cash\nBank",
            "default": "Cash",
            "reqd":0,
        },
        
	]
};



// if(frappe.query_report.get_filter_value('make_hidden') == 1){
//     $('[data-fieldname=mode_of_payment]').attr("disabled", true);
//     console.log("yeeeeeeeeeeeeees")


// }else{
//     console.log("not")

// }